### ISSUE TITLE
flask_sqlalchemy_session support?

### ISSUE THREAD

**Issue by illume** [inveniosoftware/flask-security-fork#20]
Hello,

flask_sqlalchemy_session is an alternative flask sqlalchemy module that doesn't have a tight coupling between sqlalchemy and flask. Which is nice if your code isn't using flask all the time (say you have some command line apps too).

Anyway... what do you think about adding support in for it? Below is what I'm using so far. Seems to work for me :) If you'll accept something like this, I can work up a pull request.

cheers!

```

class PretendFlaskSQLAlchemyDb(object):
    """ This is a pretend db object, so we can just pass in a session.
    """
    def __init__(self, session):
        self.session = session

class SQLAlchemyUserDatastore(SQLAlchemyDatastore, UserDatastore):
    """A SQLAlchemy datastore implementation for Flask-Security that assumes the
    use of the Flask-SQLAlchemy extension.
    """
    def __init__(self, db, user_model, role_model):
        SQLAlchemyDatastore.__init__(self, db)
        UserDatastore.__init__(self, user_model, role_model)

    def get_user(self, identifier):
        if self._is_numeric(identifier):
            return self.db.session.query(self.user_model).get(identifier)
        for attr in get_identity_attributes():
            query = getattr(self.user_model, attr).ilike(identifier)
            rv = self.db.session.query(self.user_model).filter(query).first()
            if rv is not None:
                return rv

    def _is_numeric(self, value):
        try:
            int(value)
        except (TypeError, ValueError):
            return False
        return True

    def find_user(self, **kwargs):
        return self.db.session.query(self.user_model).filter_by(**kwargs).first()

    def find_role(self, role):
        return self.db.session.query(self.user_model).filter_by(name=role).first()


class SQLAlchemySessionUserDatastore(SQLAlchemyUserDatastore):
    """A SQLAlchemy datastore implementation for Flask-Security that assumes the
       use of the flask_sqlalchemy_session extension.
    """
    def __init__(self, session, user_model, role_model):
        SQLAlchemyUserDatastore.__init__(self,
                                         PretendFlaskSQLAlchemyDb(session),
                                         user_model,
                                         role_model)

```

**Comment by jirikuncar** [comment]
It looks fine. Just make sure that you include tests and docs. Thanks!

**Comment by krahser** [comment]
Unfortunately, I couldn't figure out why the test is not passing.

`            assert datastore.add_role_to_user(user, 'editor') is True
>           assert datastore.add_role_to_user(user, 'editor') is False
E           assert True is False
`

@illume Does the method add_role_to_user work for you?

**Comment by illume** [comment]
@krahser note, the SQLAlchemyUserDatastore had some changes too. Thanks heaps for doing this!

**Comment by krahser** [comment]
Yes, I saw that, but it's not necessary. If you make the query as is defined it works too.
Just to be sure I test both ways and I get the same error.

**Comment by krahser** [comment]
I realize that the mistake was in the backref definition on the user model.

It's alredy pushed in the fix, that pass all the tests and the beginning of docs in [my fork](https://github.com/krahser/flask-security-fork/tree/flask_sqlalchemy_session)

I hope to finish this tomorrow.
ithub.com/krahser/flask-security-fork/tree/flask_sqlalchemy_session)

I hope to finish this tomorrow.

**Comment by illume** [comment]
I made some changes here: https://github.com/illume/flask-security-fork/tree/flask_sqlalchemy_session

There's still some failures with old versions of flask-sqlalchemy. http://stackoverflow.com/questions/20201809/sqlalchemy-flask-attributeerror-session-object-has-no-attribute-model-chan

I added a fix for that in the branch. All the tests pass on travis now.

**Comment by illume** [comment]
Cool. I'll leave the rest to you.  See "def commit" here: https://github.com/illume/flask-security-fork/tree/flask_sqlalchemy_session

**Comment by jirikuncar** [comment]
Closed via 731958c

**Comment by illume** [comment]
Thanks!
